Expand twister quarantine feature

nashif opened the pull request:
- twister: quarantine: use / to seperate platform from test
- twister: support quarantining platforms, architectures, simulators.
- support quarantining tests based on a regex/wildcard

nashif committed:
twister: quarantine: use / to seperate platform from test

For consistency and readability.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: support quarantining platforms, architectures, simulators

Expand quarantine to allow exclusion of plaforms, architectures and
simulators.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

PerMac reviewed (commented):
What is the rule behind "simulation" in a quarantine? From the code it looks like a platform from this section will have the execution stage turned off, i.e. became build-only? Is it so?

nashif commented:
> What is the rule behind "simulation" in a quarantine? From the code it looks like a platform from this section will have the execution stage turned off, i.e. became build-only? Is it so?

yes, here you add the type of simulation and anything platforms using this simulation type will not be executed.

PerMac reviewed (dismissed):
I like the expansion for platforms and arch but I think adding "simulation" in this form is more of a workaround/hack. I can see the need for functionality* allowing certain configurations to became build-only, however I'm worried this implementation will cause more confusion. 

The functionality for "simulation" is different than the regular workflow for quarantine, where quarantine tests are completely skipped. Also, the name is misleading. The implementation has nothing to do with simulations but rather with build-only. This can work in the same way for CIs with hw connected. We can still see image being built and getting execution skipped.

I think we can use the same file for "quarantine" and "build-only" but the difference has to be clear. E.g. we can have 2 "dictionaries". A "Quarantine" section with all the keywords for scenarios, platforms and arch - such configuration will be fully skipped; and a "Build-only" section, with the same underlining structure, from which configurations will have only execution skipped.

*however, I haven't yet experience a case where we would like to keep building a configuration, even though we know it is broken and fails to execute. Can you explain, why someone would want to do this?

nashif commented:
> *however, I haven't yet experience a case where we would like to keep building a configuration, even though we know it is broken and fails to execute. Can you explain, why someone would want to do this?

This is not about the test being broken, it is about the simulator not functioning correctly, so the test is then skipped (build only are considered skipped in the report), we can also skip completely, no problem with that.

PerMac commented on doc/develop/test/twister.rst:
```diff
@@ -888,19 +888,27 @@ be added to the output reports.
+To exclude a platform, use the following syntax::
+
+    platforms:
+      - qemu_x86
```
@nashif Please add option to add comment field to each of the new sections. It is very useful for stating the reason of quarantine. E.g. we add link to a corresponding issue and it allows for easy tracking the status of tests under quarantine.

PerMac commented:
@nashif 
>we can also skip completely, no problem with that

If so, please do that so the workflow is consistent.

nashif commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -648,9 +658,15 @@ def apply_filters(self, **kwargs):
                 if not force_platform and plat.name in exclude_platform:
                     instance.add_filter("Platform is excluded on command line.", Filters.CMD_LINE)
 
+                if plat.name in self.quarantine.platforms:
```
only tests, everything else is easily verified using other twister command line options

nashif committed:
twister: quarantine: support wildcard in scenario names

Allow wildcards in scenario names when quarantining. This will make it
easy to add all scenarios of a testsuite or just a limited set based on
a wildcard and regex matching.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

gchwier commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -471,24 +476,41 @@ def load_quarantine(self, file):
+                scenarios.append(ts.id)
+            q_scenarios = []
+            for s in quar_dict['scenarios']:
+                if '.*' in s:
```
Nice feature. But maybe instead of checking for '.*' would be better to use raw string ( r'pattern', e.g. r'kernel\.common.*')?
In current solution, if you add '.*' to tests with dots, every dot will be taken by a regex as any letter.

nashif replied:
I guess I can remove the whole thing and allow any type of regex, so this becomes:

```
+            for s in quar_dict['scenarios']:
+                r = re.compile(s)
+                q_scenarios.extend(list(filter(r.match, scenarios)))
```

gchwier commented:
@nashif Could you take a look into this PR #52470
We started implementing the quarantine feature in TwisterV2. I tried to port this here.
Should work similar to current solution. Additionally added:
* supports multiple --quarantine-list
* is consistent with previous yaml schema (instead of extra list with `platforms` an `architecture`, filtering is achieved by adding 'all' to scenarios - or just ommitting `scenarios` keyword)
* no quarantine list is generated (platform + scenario_name), which can be very large. Filtering is done on the fly.
* IMO - easier to add new features (new filter keywords, regex check etc.)
Please check it if it is worth using in Twister

teburd commented:
Makes sense to me, its nice to use regexes here to easily quarantine whole sets of things in one easy go

gchwier commented:
I found your answer under PR #52470, that "we are overloading and misusing the current syntax and scheme".
But I'm not really sure, if it is good way to add new sections and move old scheme under `tests` section.
In Downstream CI jobs are used quarantine yaml files and after this change all quarantine files must be updated.

I updated PR #52470, added `simulation` and regex (not only to `scenarios`, but also to other fields).
Example quarantine file could be like that:
```
- scenarios:
  - kernel.common.(misra|tls)
  comment: "link to issue"

# quarantine kernel.common scenario for all platforms with name started from nrf53
- scenarios:
  - kernel.common
  platforms:
  - nrf53.*

# quarantine all scenarios where simulation is qemu, without reason
- simulations:
  - qemu
```

gchwier commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -662,9 +682,17 @@ def apply_filters(self, **kwargs):
                             if ts.harness_config.get('fixture') in h.fixtures:
                                 instance.run = True
 
+                if plat.simulation in [q['name'] for q in self.quarantine.simulation]:
```
This method (`apply_filters`) is really big and not easy to read.
If there is separate Quarantine class, some methods could be added there (_search_comment, is_platform_under_quarantine, load_quarantinte_from_yaml etc.)

gchwier commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -473,24 +478,37 @@ def load_quarantine(self, file):
+            q_scenarios = []
+            for s in quar_dict['scenarios']:
+                r = re.compile(s)
+                q_scenarios.extend(list(filter(r.fullmatch, scenarios)))
```
Must be aware, that every dot in scenario (`kernel.common.misra`) will be treated as 'any char'. Probably there is no scenarios, where it does matter.

gchwier commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -473,24 +478,37 @@ def load_quarantine(self, file):
         # Change quarantine_list into a dictionary
         for d in quarantine_list:
-            self.quarantine.update(d)
+            self.quarantine.tests.update(d)
```
We store here a list of all possible configurations to be quarantined (a Cartesian product of scenarios and platform), the list can be very large. Maybe it could be enough to keep in Quarantine class only data from yaml file and do filtering in `apply_filters` method (like for newly added platforms, architectures ...).
